contracts/StablecoinConverter.sol

```diff
@@ -0,0 +1,119 @@
+pragma solidity ^0.5.0;
+
+import "./EpochTokenLocker.sol";
+import "./libraries/IdToAddressBiMap.sol";
+import "solidity-bytes-utils/contracts/BytesLib.sol";
+import "./libraries/IdToAddressBiMap.sol";
+
+
+contract StablecoinConverter is EpochTokenLocker {
+    using SafeMath for uint;
+    using BytesLib for bytes32;
+
+    event OrderPlacement(
+        address owner,
+        uint16 buyToken,
+        uint16 sellToken,
+        bool sellOrderFlag,
+        uint32 validFrom,
+        uint32 validUntil,
+        uint256 buyAmount,
+        uint256 sellAmount
+    );
+
+    event OrderCancelation(
+        address owner,
+        uint id
+    );
+
+    //outstanding volume of an order is encoded as buyAmount or sellAmount depending on sellOrderFlag
+    struct Order {
+        uint16 buyToken;
+        uint16 sellToken;
+        uint32 validFrom;  // order is valid from validFrom inclusively
+        uint32 validUntil;  // order is valid till validUntil inclusively
+        bool sellOrderFlag;
+        uint128 buyAmount;
+        uint128 sellAmount;
+    }
+
+    // User-> Order
+    mapping(address => Order[]) public orders;
+
+    IdToAddressBiMap.Data private registeredTokens;
+    uint public MAX_TOKENS;
+    uint16 public numTokens = 0;
+
+    constructor(uint maxTokens) public {
+        MAX_TOKENS = maxTokens;
+    }
+
+    function addToken(address _tokenAddress) public {
+        require(numTokens < MAX_TOKENS, "Max tokens reached");
+        require(
+            IdToAddressBiMap.insert(registeredTokens, numTokens, _tokenAddress),
+            "Token already registered"
+        );
+        numTokens++;
+    }
+
+    function placeOrder(
+        uint16 buyToken,
+        uint16 sellToken,
+        bool sellOrderFlag,
+        uint32 validUntil,
+        uint128 buyAmount,
+        uint128 sellAmount
+    ) public returns (uint) {
+        orders[msg.sender].push(Order({
+            buyToken: buyToken,
+            sellToken: sellToken,
+            validFrom: getCurrentStateIndex() + 1,
+            validUntil: validUntil,
+            sellOrderFlag: sellOrderFlag,
+            buyAmount: buyAmount,
+            sellAmount: sellAmount
+        }));
+        emit OrderPlacement(
+            msg.sender,
+            buyToken,
+            sellToken,
+            sellOrderFlag,
+            getCurrentStateIndex() + 1,
+            validUntil,
+            buyAmount,
+            sellAmount
+        );
+        return orders[msg.sender].length - 1;
+    }
+
+    function cancelOrder(
+        uint id
+    ) public {
+        orders[msg.sender][id].validUntil = getCurrentStateIndex();
+        emit OrderCancelation(msg.sender, id);
```

Comment:
I think this is good for the sake of simplicity, but would it be worth it to do an actual delete of the order?

Also, shouldn't this function perform some checks that the order hasn't been filled?

Reply:
NVM: I just saw `freeStorageOfOrder`. But why don't we call it when we do the cancelation already?

Also `orders[msg.sender][id].validUntil = getCurrentStateIndex();` states that in the current state index is valid. Someone can still fill our order, even after the user call `cancelOrder`.

Reply:
Good catch. I think this should be `getCurrentStateIndex() -1`, since validUntil is inclusive. We want to cancel it for any batches for which price finding has not yet started (since otherwise the solver might have to start from scratch again). However, I believe the solver is always running for `getCurrentStateIndex() -1`.